content/blog/2019-09-25-c-implementation.md

```diff
@@ -0,0 +1,189 @@
+right place (or in the case of effect operations, has teh correct effect). 
+
+## Op Code Implementation
+All instructions but `jmp`, `br`, `print`, and `const` have a trivial
+implementation. A function called `get_or_quit` fetches arguments from the
+source program and stores their values in a global array, sized to the max of
+the argument count of all non-print instructions (2). If the variable is not
+found in storage, an error is reported (containing the specific issue,
+incorrect variable, and instruction pointer value) and the program quits.
+Otherwise, the op implementing function calls `put` with the argument that
+implements the specific op code, which stores the result in the destination
+variable.. For example, the `add` op code is implemented
+as follows:
+
+```c
+static void op_add() {
+  get_or_quit();
+  put(mem_args[0]+mem_args[1]);
+}
+```
+Above, `mem_args` is the name of the global argument array. Thus, most op codes
+have 2 line implementations. Even `br`, the most complex op code, requires only
+16 lines to implement. 
+
+## Memory Implentation (and More, for Free)
+I needed a hash table implementation to implement variable storage, so I built
+it myself. It can be found in `src/table.{c,h}`. The dictionary uses open
+addressing and linear chaining. The hash function copies the hash used by
+java.lang.String’s hashCode() method. The `_hash` function computes
+a polynomial whos coeffecients are the integer values of the string's
+characters, evaluated at `x=31`. The polynomial is evaluated with Horner's
+method. The `table` maps string keys to `int64_t` values.  I represent bril
+integers and bril booleans with `int64_t` to avoid storing the type of bril
+variables. Since bril's typesystem only allows for those two types, this design
+is sufficient for now. Incorporating more types will be simple: I can extend
+the `table_elem` struct with a type bitfield/enum value.  For now, this simple
+table suffices. 
+
+The table code is also used for the label->index map and to store the mapping
+between string op codes and the index for that op code's implementing function
+in an array of function pointers.
+
+## Evaluation
+### Correctness
+I ran `cril` against the programs in `bril/test` and made sure the outputs
+matched `<program>.out`. My interpreter gave correct results on a fibonacci
+program posted in the class slack channel, on which the reference interpreter
+gave incorrect results. 
+
+### Performance
+I wanted to have a way to get reliable performance numbers to square off
+againts any other student's implementation of bril. First, I had the main
+evaluation loop return a `uint64_t` number of nanoseconds of elapsed time.
+I used the POSIX provided `timespec` structure to record the time tracked by
+`CLOCK_PROCESS_CPUTIME_ID`, the nanosecond resolution process time clock. This
+tracks CPU ticks spent on the program process itself, irrespective of other
+scheduled processes. The code is reproduced below:
```

Comment:
I don't think this blob of code adds too much to your discussion—I think it suffices to discuss where your clock came from, which is libc's `clock_gettime`, and its resolution. (The bullet points below the code, for example, are plenty of detail.)